dictionary.go

```diff
@@ -141,69 +164,88 @@ func (d *byteArrayDictionary) Bounds(indexes []int32) (min, max Value) {
 	return min, max
 }
 
-func (d *byteArrayDictionary) ReadFrom(decoder encoding.Decoder) error {
-	d.Reset()
-	for {
-		if d.values.Len() == d.values.Cap() {
-			d.values.Grow(d.values.Len())
-		}
-		_, err := decoder.DecodeByteArray(&d.values)
-		if err != nil {
-			if err == io.EOF {
-				err = nil
-			}
-			return err
-		}
-	}
-}
-
-func (d *byteArrayDictionary) WriteTo(encoder encoding.Encoder) error {
-	if err := encoder.EncodeByteArray(d.values); err != nil {
-		return fmt.Errorf("writing parquet dictionary of %d binary values: %w", d.Len(), err)
-	}
-	return nil
-}
-
 func (d *byteArrayDictionary) Reset() {
 	d.values.Reset()
 	d.index = nil
 }
 
+func (d *byteArrayDictionary) Page() BufferedPage {
+	return &d.byteArrayPage
+}
+
 type fixedLenByteArrayDictionary struct {
-	typ    Type
-	size   int
-	values []byte
-	index  map[string]int32
+	fixedLenByteArrayPage
+	typ   Type
+	index map[string]int32
 }
 
-func newFixedLenByteArrayDictionary(typ Type, bufferSize int) *fixedLenByteArrayDictionary {
+func newFixedLenByteArrayDictionary(typ Type, columnIndex int16, bufferSize int) *fixedLenByteArrayDictionary {
 	size := typ.Length()
 	return &fixedLenByteArrayDictionary{
-		typ:    typ,
-		size:   size,
-		values: make([]byte, 0, dictCap(bufferSize, size)*size),
+		typ: typ,
+		fixedLenByteArrayPage: fixedLenByteArrayPage{
+			size:        size,
+			data:        make([]byte, 0, dictCap(bufferSize, size)*size),
+			columnIndex: columnIndex,
+		},
 	}
 }
 
+func readFixedLenByteArrayDictionary(typ Type, columnIndex int16, numValues int, decoder encoding.Decoder) (Dictionary, error) {
+	size := typ.Length()
+
+	d := &fixedLenByteArrayDictionary{
+		typ: typ,
+		fixedLenByteArrayPage: fixedLenByteArrayPage{
+			size:        size,
+			data:        make([]byte, 0, atLeastOne(numValues)*size),
+			columnIndex: columnIndex,
+		},
+	}
+
+	for {
+		if len(d.data) == cap(d.data) {
+			newValues := make([]byte, len(d.data), 2*cap(d.data))
+			copy(newValues, d.data)
+			d.data = newValues
+		}
+
+		n, err := decoder.DecodeFixedLenByteArray(d.size, d.data[len(d.data):cap(d.data)])
+		if n > 0 {
+			d.data = d.data[:len(d.data)+(n*d.size)]
+		}
+
+		if err != nil {
```

Comment:
This reads a little awkward... to mix the success and the error cases together in the same if statement, what about

```go
if err == io.EOF {
    return d, nil
}
if err != nil {
    return d, fmt.Errorf(...)
}
```